Source/DafnyLanguageServer/README.md

```diff
@@ -121,4 +130,122 @@ Plugins are typically used to report additional diagnostics such as unsupported
 
 Note that all plugin errors should use the original program's expressions' token and NOT `Token.NoToken`, else no error will be displayed in the IDE.
 
-Morover, plugins should not write anything to `stdout` as it interferes with the communication protocol with the IDE.
\ No newline at end of file
+## Code actions plugin tutorial
+
+In this section, we will create a plugin to provide more code actions to Dafny.
+The code actions will be simple: Add a dummy comment in front of the first method name,
+if the selection is on the line of the method.
+
+Assuming the Dafny source code is installed in the folder `dafny/`
+start by creating an empty folder next to it, e.g. `PluginTutorial/`
+
+```
+> mkdir PluginTutorial
+> cd PluginTutorial
+```
+Then, create a dotnet class project
+```
+> dotnet new classlib
+```
+It will create a file `Class1.cs` that you can rename
+```
+> mv Class1.cs PluginAddComment.cs
+```
+Open the newly created file `PluginTutorial.csproj`, and add the following after `</PropertyGroup>`:
+```
+  <ItemGroup>
+    <ProjectReference Include="../dafny/source/DafnyLanguageServer/DafnyLanguageServer.csproj" />
+  </ItemGroup>
+```
+
+
+Then, open the file `PluginAddComment.cs`, remove everything, and write the imports and a namespace:
+
+```csharp
+using Microsoft.Dafny;
+using Microsoft.Dafny.LanguageServer.Plugins;
+using Microsoft.Boogie;
+using Microsoft.Dafny.LanguageServer.Language;
+using Range = OmniSharp.Extensions.LanguageServer.Protocol.Models.Range;
+
+namespace PluginAddComment;
+```
+
+After that, add a `PluginConfiguration` that will expose all the quickfixers of your plugin.
+This class will be discovered and instantiated automatically by Dafny.
+```
+public class TestConfiguration : PluginConfiguration {
+  public override QuickFixer[] GetQuickFixers() {
+    return new QuickFixer[] { new AddCommentQuickFixer() };
+  }
+}
+```
+Note that you could also override the methods `GetRewriters()` and `GetCompilers()` for other purposes, but this is out of scope for this tutorial.
+
+Then, we need to create the quickFixer `AddCommentQuickFixer` itself:
+
+```
+public class AddCommentQuickFixer : QuickFixer {
+  public override QuickFix[] GetQuickFixes(IQuickFixInput input, Range selection) {
```

Comment:
I hear you, but that means it requires clients to include Ling in their plugin, which they don't have. Clients should be able to write plugins using languages that's don't have IEnumerable (i.e. Dafny ...)

Reply:
This is a C# API, so it should be idiomatic to C#.

If you want to have a Dafny API for a piece of the Dafny compiler written in C#, then you should add a Dafny layer on top of the C# API.

Reply:
See this discussion here we had previously in the context of plugins for the first reason why it's not a good idea to use C# idioms.
https://github.com/dafny-lang/dafny/pull/1739#discussion_r802868967
In short, it is not meant to be a C# API. It's a DotNet API. And it's not only Dafny that will use it, but also F#.

The second reason I'll make is that we need all the summary of these quick fixes immediately. An `IEnumerable` would make sense if we were streaming them or if it was expensive to compute all of them. But since they are going to appear in the UI of the user, there won't be more than 10 of these items usually.
You can always use an array as an `IEnumerable` by including Linq.

Reply:
> In short, it is not meant to be a C# API. It's a DotNet API. And it's not only Dafny that will use it, but also F#.

Does C# vs F# change anything about IEnumerable being idiomatic?

> The second reason I'll make is that we need all the summary of these quick fixes immediately

You could argue for returning an IReadOnlyList, but I think returning an abstract type like `IEnumerable` or `IReadOnlyList` instead of a specific type like an array is still idiomatic C#/.NET here.

Reply:
I forgot to reply, thanks for your patience and review @robin-aws  and @keyboardDrummer 

I disagree that IEnumerable (🥈) is "best" for an API than a regular array (🥇) , because it requires to import `use System.Collection.Generic`, enumerables take more space on the disk, and in other parts of the API (`GetRewriters`, `GetCompilers`, `GetQuickFixers`), where we don't need anything enumerable, we are using arrays.

However, given that you were 2️⃣  to ask for it, and it does not make that much difference at the end, and it's not at the same level as the top-level `GetRewriters` ..., I did implement the change to use `IEnumerable<>` instead of an array 🍨 . Let me know if you like it.